cmd/asset-syncer/utils.go

```diff
@@ -431,77 +445,127 @@ func extractFilesFromBuffer(buf *bytes.Buffer) (*artifactFiles, error) {
+		Maintainers:   maintainers,
+		Sources:       chartMetadata.Sources,
+		Icon:          chartMetadata.Icon,
+		Category:      chartMetadata.Annotations["category"],
+		ChartVersions: []models.ChartVersion{chartVersion},
+	}, nil
+}
+
+func chartImportWorker(repoURL *url.URL, r *OCIRegistry, chartJobs <-chan pullChartJob, resultChan chan pullChartResult) {
+	for j := range chartJobs {
+		log.WithFields(log.Fields{"name": j.AppName, "tag": j.Tag}).Debug("pulling chart")
+		chart, err := pullAndExtract(repoURL, j.AppName, j.Tag, r.puller, r)
+		resultChan <- pullChartResult{chart, err}
+	}
+}
+
 // Charts retrieve the list of charts exposed in the repo
 func (r *OCIRegistry) Charts() ([]models.Chart, error) {
-	result := []models.Chart{}
+	result := map[string]models.Chart{}
 	url, err := parseRepoURL(r.RepoInternal.URL)
 	if err != nil {
 		return nil, err
 	}
 
+	chartJobs := make(chan pullChartJob, numWorkers)
+	chartChan := make(chan pullChartResult)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	totalCharts := 0
 	for _, appName := range r.repositories {
-		chartInfo := models.Chart{}
-		firstTag := true
-		for _, tag := range r.tags[appName].Tags {
-			// Pull Chart
-			// TODO: Run in parallel
-			ref := path.Join(url.Host, url.Path, fmt.Sprintf("%s:%s", appName, tag))
-			chartBuffer, digest, err := r.puller.PullOCIChart(ref)
-			if err != nil {
-				return nil, err
-			}
-
-			// Extract
-			files, err := extractFilesFromBuffer(chartBuffer)
-			if err != nil {
-				return nil, err
-			}
-			chartMetadata := h3chart.Metadata{}
-			err = yaml.Unmarshal([]byte(files.Metadata), &chartMetadata)
-			if err != nil {
-				return nil, err
-			}
-
-			// Format Data
-			chartVersion := models.ChartVersion{
-				Version:    chartMetadata.Version,
-				AppVersion: chartMetadata.AppVersion,
-				Digest:     digest,
-				URLs:       chartMetadata.Sources,
-				Readme:     files.Readme,
-				Values:     files.Values,
-				Schema:     files.Schema,
-			}
-			if firstTag {
-				firstTag = false
-				maintainers := []chart.Maintainer{}
-				for _, m := range chartMetadata.Maintainers {
-					maintainers = append(maintainers, chart.Maintainer{
-						Name:  m.Name,
-						Email: m.Email,
-						Url:   m.URL,
-					})
-				}
-				chartInfo = models.Chart{
-					ID:            path.Join(r.Name, appName),
-					Name:          appName,
-					Repo:          &models.Repo{Namespace: r.Namespace, Name: r.Name, URL: r.URL, Type: r.Type},
-					Description:   chartMetadata.Description,
-					Home:          chartMetadata.Home,
-					Keywords:      chartMetadata.Keywords,
-					Maintainers:   maintainers,
-					Sources:       chartMetadata.Sources,
-					Icon:          chartMetadata.Icon,
-					Category:      chartMetadata.Annotations["category"],
-					ChartVersions: []models.ChartVersion{chartVersion},
+		totalCharts += len(r.tags[appName].Tags)
+	}
+	// Process 10 charts at a time
+	for i := 0; i < numWorkers; i++ {
+		go chartImportWorker(url, r, chartJobs, chartChan)
+	}
```

Comment:
Replace this loop with something like:
```
	// Process 10 charts at a time ensuring that each worker signals when complete.
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			chartImportWorker(url, r, chartJobs, chartChan)
			wg.Done()
		}()
	}
	// When we know all workers have sent their data in chartChan, close it.
	go func() {
		wg.Wait()
		chartChan.Close()
	}
```

Reply:
closing the results channel did the trick :)